### PR TITLE
Call Dispose in Finalizer; remove unnecessary Null check

### PR DESCRIPTION
**Pull request by jbe2277**
Please compare with the original implementation which behaves identical to my changes: http://referencesource.microsoft.com/#System/compmod/system/componentmodel/BackgroundWorker.cs,85d60b0d93a826fa

### CODE REVIEW

**Comment by stephentoub** [comment]
Nice catch on the finalizer. LGTM.

**Comment by mikedn** [comment]
This type should not have a finalizer, it doesn't hold any unmanaged resources.

**Comment by stephentoub** [comment]
@mikedn, it's unsealed, exposes a virtual Dispose method, and its implementation in the .NET Framework has a finalizer method that calls Dispose(false).  That means someone who derives from BackgroundWorker could end up with different behavior depending on whether they're running on the full framework or .NET Core.  If we were to say that BackgroundWorker in .NET Core is sealed and/or doesn't provide a virtual Dispose(bool), then I'd be fine with it not having the matching finalizer. @davkean?

**Comment by mikedn** [comment]
I suppose that there's a small chance to break some strange code (I've never seen anyone inheriting from BackgroundWorker. The chance that such a derived type holds unmanaged resources is even smaller). But let's look at the big picture:
* BackgroundWorker is primarily used in WinForms apps. WinForms apps can't run on .NET Core anyway.
* The WinStore profile doesn't even include this type.
* Silverlight's BackgroundWorker isn't disposable.

Oh well, I suppose it doesn't really matter if it's finanlizable in .NET Core. No one will use this type anyway :smile:.

**Comment by sharwell** [comment]
@stephentoub That is not the correct rule for determining when a user-defined finalizer should be provided. A user-defined finalizer should only be provided if the type directly owns unmanaged resources. If a type extends `BackgroundWorker` and introduces its own unmanaged resources, that type is responsible for including a finalizer.

Regarding the change itself:

:exclamation: This pull request introduces an unnecessary finalizer, which increases overhead for any code using the object. **However**, the .NET Framework implementation of the same class *does* include a finalizer (provided by the `Component` class), so omitting the finalizer here could alter the behavior for certain code that meets all of the following conditions:

1. A type is derived from `BackgroundWorker`
2. This derived type directly owns unmanaged resources (as opposed to wrapping them in a `SafeHandle` or similar)
3. The consumer of this type is not calling the implementation of `IDisposable.Dispose()` to release resources

Note that a type like the above could define its own finalizer to ensure `Dispose(false)` is called if necessary, but this type would then make *two* calls to `Dispose(false)` when used with the .NET Framework. The recommended solution is to always wrap unmanaged resources in a dedicated holder like `SafeHandle`, but obviously not everyone does this.

If this list of considerations meets the bar for preserving compatibility across builds, then the finalizer should be included. Otherwise, I would consider leaving the finalizer out to avoid unnecessary performance overhead.

**Comment by stephentoub** [comment]
> That is not the correct rule for determining when a user-defined finalizer should be provided

Not sure why you say I'm incorrect... my description included the compat risk of the framework's version having a finalizer: "and its implementation in the .NET Framework has a finalizer method that calls Dispose(false)".

**Comment by sharwell** [comment]
> Not sure why you say I'm incorrect...

I misread what you wrote, and now removed that comment. I believe we are otherwise in agreement regarding the analysis of the situation.

**Comment by weshaggard** [comment]
This is an interesting case but I'm kind of leaning towards not adding this finializer, because I feel the risk isn't high enough to warrant the finializer (I really dislike finializers :)). While the analysis is correct and there is potential to leak some unmanaged resources in some corner cases where the Dispose method in a derived type isn't ever called. I think we have similar issues in other places and we haven't tried to fill the gap. You can imagine any method on a type that we removed/filtered out in .NET Core but exists on full .NET Framework (ex: various Close methods) could cause similar issues.

So I would honestly not add this finializer unless we have known breaks.

**Comment by stephentoub** [comment]
I don't feel strongly about it, but I lean the other way.  

As I see it, BackgroundWorker exists in .NET Core solely for compat reasons; there are better options available now for accomplishing the same thing with less code, with cleaner code, and more efficiently.  And as such we should strive to make it as compatible as possible, such that the folks who will be using it (those that already have code written to use it) have the best chance of not having their code break in mysterious ways.  While I agree this particular case should be rare, it'd also be very difficult for someone in this situation to diagnose, likely only showing up in production, as slowdowns or hangs or worker reboots and the like.

If we decide not to add back the finalizer, why have Dispose(bool) call GC.SuppressFinalize?  For that matter, why have it implement IDisposable at all?

**Comment by jbe2277** [comment]
I agree. If we intend to keep the breaking change with the missing Dispose call in the finalizer then please do it in a way that we get immediate feedback for breaking legacy code (e.g. compiler error). This could be done by marking the class as _sealed_ or removing the _Dispose_ implementation from the class.

**Comment by sharwell** [comment]
The current implementation:

* Works for users who add only managed resources, which includes adding managed wrappers for unmanaged resources. This is the recommended practice for all development involving unmanaged resources.
* Works for users who properly dispose of objects after they are done using them, even if they directly added unmanaged resources.

Removing the call to `GC.SuppressFinalize(this)` is not recommended, even if the type does not define a user-defined finalizer. The call to `GC.SuppressFinalize` appears in the implementation of `IDisposable.Dispose()`, which is `sealed`. If a derived type adds a finalizer, that type ~~cannot~~ should not also override the implementation of `IDisposable.Dispose()`, but still expects `GC.SuppressFinalize(this)` to be called after the `Dispose()` operation completes. Placing the call to `GC.SuppressFinalize(this)` within an override of `Dispose(bool)` is also not recommended per the ordering discussion in [this article](http://joeduffyblog.com/2005/04/08/dg-update-dispose-finalization-and-resource-management/).

**Comment by sharwell** [comment]
Whether or not it's needed, removing the implementation of `IDisposable` has a *very* high likelihood of breaking users due to [CA2213](https://msdn.microsoft.com/en-us/library/ms182328.aspx) reminding users to actually call this method.

**Comment by sharwell** [comment]
Marking the type `sealed` or the `Dispose(bool)` method `sealed` could cause a compatibility problem for people correctly using the type, which is obviously undesirable.

I'll change my recommendation here to be **in favor of adding the finalizer**, provided *either* of the following conditions holds:

1. The `BackgroundWorker` type is being included for compatibility purposes only (similar to the non-generic collection types), and is not considered to be the right option for any new development.
2. The majority of use cases involving `BackgroundWorker` use the object as a long-lived object that is infrequently instantiated and executes many "operations", as opposed to a type like `Task` which gets instantiated perhaps once per "operation".

**Comment by stephentoub** [comment]
Both of those reasons are why I've been saying we should just add the finalizer.

Regarding, the .NET Core surface area is already smaller than what developers on the .NET Framework have available to them.  There are plenty of cases where code written for the .NET Framework will not compile when targeting .NET Core.  Making the type sealed for example would just be one more example of that.  The type already lost its base class in .NET Core, which means it lost other APIs, like the Disposed event.

I would much rather have an app fail at compile-time due to decreased API surface area than have an app mysteriously fail at run-time due to lack of a finalizer.  And I'm extremely doubtful that the finalizer will be problematic for performance on this type... even if it is, it's already an issue in apps using BackgroundWorker because the BackgroundWorker that's been shipping has a finalizer, so it's not making things worse.

So, I still believe we should either a) add back the finalizer as this PR proposes, or b) seal the type and/or remove its IDisposable implementation.

**Comment by sharwell** [comment]
> Both of those reasons are why I've been saying we should just add the finalizer.

I'm a "tough sell", which means in addition to taking that side you explained your points well.¹

It wouldn't hurt to have additional people agreeing the specific item(s) I mentioned hold for this type.

¹ My hope is that this is a logical conclusion, but... :wink:

**Comment by mikedn** [comment]
I'd remove the IDisposable implementation. BackgroundWorker never needed it, it inherited it from Component. BackgroundWorker was a Component because you were supposed to be able to add it to a WinForms Form/UserControl.

I'm not sure what to do about the analysis warning mentioned by @sharwell. I'm tempted to say that it's not a real problem because most people will probably use PCLs if they want to target both .NET Framework and .NET Core. In that case Dispose will probably not show up in the contract so you shouldn't see CA2213.

File: src/System.ComponentModel.EventBasedAsync/src/System/ComponentModel/BackgroundWorker.cs [hunk 1]
```diff
@@ -23,6 +23,11 @@ public BackgroundWorker()
             _operationCompleted = new SendOrPostCallback(AsyncOperationCompleted);
             _progressReporter = new SendOrPostCallback(ProgressReporter);
         }
+        
+        ~BackgroundWorker()
```
Comment (sharwell) [inline src/System.ComponentModel.EventBasedAsync/src/System/ComponentModel/BackgroundWorker.cs]: Can you add a comment here explaining that this finalizer exists for compatibility purposes, with a link to this pull request?

**Comment by sharwell** [comment]
> I'm not sure what to do about the analysis warning mentioned by @sharwell.

My specific concern is the increased work involved with porting certain code from the desktop framework to .NET Core. If the desktop users ran code analysis in the past, it may have instructed them to actually call `Dispose()` on `BackgroundWorker` objects (or place them in a `using` statement), even if it wasn't needed. If the `IDisposable` implementation is removed from .NET Core, then not only will the code fail to compile on .NET Core but that code will now produce the code analysis warning again if/when it is shared with desktop code.

**Comment by davkean** [comment]
When we originally reviewed the proposed surface area of BackgroundWorker, we were under the impression that Dispose actually did something [1] - and hence the advise was to include it. Given that's not the case, _I would agree that we should just remove it's IDisposable implementation_.

@sharwell Yes code ported from desktop will fail to compile, but so will code that hooked onto the Disposed event, or derived from the BackgroundWorker and used the Events property (which has twice as much usage[1] as calling Dispose). I'm not concerned at all by customers copy and pasting code from .NET Core -> Desktop and getting an FxCop warning.

[1] We expected it to at least stop firing events
[2] I pulled the .NET Framework data we have, and using the Events property from Component has nearly twice as much usage as calling Dispose.

**Comment by nguerrera** [comment]
I think we should just match the desktop and add the finalizer. This class exists in .NET Core for compatibility alone, so let's just make it as compatible as possible.

On the separate general question of whether an unsealed IDiposable class without a finalizer should call GC.SuppressFinalize(this), @sharwell is right that it should. A derived class can then add a finalizer if needed that calls Dispose(false) and override Dispose(bool). We decided on that as a pattern long ago. See [TextWriter](https://github.com/dotnet/coreclr/blob/ef1e2ab328087c61a6878c1e84f4fc5d710aebce/src/mscorlib/src/System/IO/TextWriter.cs#L131) and [Stream](https://github.com/dotnet/coreclr/blob/ef1e2ab328087c61a6878c1e84f4fc5d710aebce/src/mscorlib/src/System/IO/Stream.cs#L242)

**Comment by mikedn** [comment]
> My specific concern is the increased work involved with porting certain code from the desktop framework to .NET Core.

I understand that but let's be realistic, how much code do you think there's out there that uses BackgroundWorker, calls its Dispose method and can be ported to .NET Core? Make sure you read my second post in this PR before answering.

The more I think about this the more I wonder why do we even have BackgroundWorker in .NET Core given that its typical usage scenario (WinForms, WPF, Silverlight) isn't valid in .NET Core and that a lot of other things from System.ComponentModel were thrown out (for good reasons).

**Comment by nguerrera** [comment]
Missed @davkean's reply before mine.

Can we not add back an Events property as well rather than arguing that it's not compatible with even more common patterns? I'm concerned about even having BackgroundWorker in .NET Core if it's not even compatible. Wasn't that the whole point -- just like non-generic collections?

**Comment by stephentoub** [comment]
@terrajobst, how would you like to proceed with this PR?

**Comment by jasonwilliams200OK** [comment]
> how would you like to proceed with this PR?

IMO by pressing the merge button. Later it can be decided in a separate issue whether to tank `BackgrounWorker` form corefx as @mikedn suggested or unimplemented `IDisposable`.

**Comment by stephentoub** [comment]
@terrajobst, can we discuss this at the next API meeting?

**Comment by stephentoub** [comment]
Let's split this into two things:
- We've stated that BackgroundWorker in .NET Core exists purely for backwards compatibility and making it possible to port existing code that uses it, we should be as true to that as possible and not inadvertently remove the finalizer.  Its use will be very limited, performance will not be the driving factor in its design, we'll actively push to migrate folks away from it, compatibility is king for this type, and so we effectively merge this PR to put back the missing finalizer.
- Then subsequently we can hash out whether there's value in having BackgroundWorker in .NET Core at all, and if that value is purely around compatibility.  Possible subsequent actions could involve saying that there's value in BackgroundWorker on its own (unlikely), in which case maybe we say compatibility isn't paramount, and we remove the IDisposable implementation (and finalizer).  Or maybe we say that BackgroundWorker doesn't serve much purpose even for compatibility and we remove it entirely.

**Comment by stephentoub** [comment]
Replaced by #2827

**Pull request closed without merging**